Deprecation of T**DetId in CalibTracker

alesaggio opened the pull request:
In the modified files we don't have a way to access the edm::EventSetup, hence the TrackerTopology cannot be retrieved. The only solution is to get specific pieces of code from T**DetId.h (in such a way that it's not easy to use from elsewhere).

cc @pieterdavid @OlivierBondu @vidalm

alesaggio committed:
Deprecation of T**DetId in CalibTracker

SiStripLorentzAngle and SiStripChannelGain

cmsbuild commented:
A new Pull Request was created by  @alesaggio  for master.

It involves the following packages:

CalibTracker/SiStripChannelGain
CalibTracker/SiStripLorentzAngle


@ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @gbenelli, @tocheng, @jlagram, @OlivierBondu, @mmusich this is something you requested to watch as well.
@davidlange6 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mmusich commented on CalibTracker/SiStripChannelGain/test/UsefullCode/TimeHistory/PlotMacro.C:
```diff
@@ -19,13 +17,27 @@
+uint32_t getTIDwheel     ( DetId detId ) { return ((detId.rawId()>>11) & 0x3); };
+uint32_t getTECside      ( DetId detId ) { return ((detId.rawId()>>18) & 0x3); };
+uint32_t getTECring      ( DetId detId ) { return ((detId.rawId()>> 5) & 0x7); };
+uint32_t getTECwheel     ( DetId detId ) { return ((detId.rawId()>>14) & 0xF); };
```
@mdelcourt I think you are the main user for this code. 
Can you just have a cross-check look? Thanks.

mmusich commented:
tracked at https://github.com/cms-sw/cmssw/issues/19398

delaere commented:
I would suggest to take a different approach for these macros: instead of recoding the detid methods (or TrackerTopology methods), you could instantiate your own TrackerTopology. The only thing you have to do is to hardcode one instance of the TOBValues struct (and others). I think that it makes the code more readable and easy to change since all hardcoded values are there and not in the various methods.

It makes the code more similar to the rest of CMSSW, and moves the burden/risk to the definition of these structs, in sync with the geometry.

Maybe (later?) this could even be read from trackerParameters.xml

davidlange6 commented:
eg a new constructor to TrackerTopology

TrackerTopology::TrackerTopology(enum LegacyTrackerNumbering)

it could have a rather big impact on any code called from a tight loop.

davidlange6 commented:
but at least centralizes all the hardwired numbers that will eventually be wrong

alesaggio committed:
Adding LegacyTrackerTopology header

alesaggio committed:
Using LegacyTrackerTopology in CalibTracker/SiStripLorentzAngle/plugins/MeasureLA.cc and CalibTracker/SiStripChannelGain/test/UsefullCode/TimeHistory/PlotMacro.C

cmsbuild commented:
Pull request #19393 was updated. @perrotta, @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @slava77, @davidlange6, @lpernie can you please check and sign again.

alesaggio commented:
I added a header file (LegacyTrackerTopology.h) that hardcodes TrackerTopology instances as discussed above. Please let me know what you think

mmusich commented on CalibTracker/SiStripLorentzAngle/plugins/MeasureLA.cc:
```diff
@@ -1,10 +1,17 @@
 #include <TChain.h>
 #include <TFile.h>
 
+namespace {
```
@alesaggio is this still needed after having introduced the LegacyTopology stuff?

alesaggio replied:
No, indeed! I'll fix it

mmusich commented on DataFormats/TrackerCommon/interface/LegacyTrackerTopology.h:
```diff
@@ -0,0 +1,97 @@
+
+namespace LegacyTrackerTopology {
+  
+  inline static std::unique_ptr<TrackerTopology> getTrackerTopology()
```
Hi @alesaggio, 
I think there might be a bit of ambiguity on what "Legacy" actually means in this context for what concern the Pixel part. Your implementation here is for Phase-0 Pixel Tracker only - I think it should be written somewhere as a disclaimer or allow the possibility to construct the Topology in this way also for Phase-I detector (since Phase0 Strips overlap in time with Phase-I Pixel, i.e. the current detector)

mmusich commented on CalibTracker/SiStripLorentzAngle/src/LA_Filler.cc:
```diff
@@ -85,11 +86,14 @@ granularity(const SiStripDetId detid, const float tthetaL, const Long64_t TFE_in
+  unsigned layer = isTIB ? tTopo_->tibLayer(detid) : tTopo_->tobLayer(detid);
+  bool stereo = isTIB ? tTopo_->tibStereo(detid) : tTopo_->tobStereo(detid);
+  
+  return subdetLabel(detid)+"_layer"+boost::lexical_cast<std::string>(layer)+(stereo?"s":"a");
```
in light of  #19645 I guess you can get rid of the ugly `boost::lexical_cast` in favor of `std::to_string`

alesaggio committed:
removing getTIBOrTOBLayer() and getTIBorTOBStereo() functions

cmsbuild commented:
Pull request #19393 was updated. @perrotta, @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @slava77, @davidlange6, @lpernie can you please check and sign again.

cmsbuild commented:
Pull request #19393 was updated. @perrotta, @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @slava77, @davidlange6, @lpernie can you please check and sign again.

alesaggio committed:
adding '#include <memory>' in LegacyTrackerTopology.h

cmsbuild commented:
Pull request #19393 was updated. @perrotta, @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @slava77, @davidlange6, @lpernie can you please check and sign again.

slava77 commented on DataFormats/TrackerCommon/interface/LegacyTrackerTopology.h:
```diff
@@ -0,0 +1,98 @@
+#include <memory>
+
+// WARNING: this header has been introduced to call a TrackerTopology
+// object whenever it is not possible to access it through an EventSetup.
```
what are the cases where the EventSetup can not deliver this?
TrackerTopologyEP and relevant upstream can be set up to provide the phase-0 topology just as it is done in normal running.

mmusich replied:
Hi @slava77, a couple of use cases are:
   * from within a ROOT macro (as in  `CalibTracker/SiStripChannelGain/test/UsefullCode/TimeHistory/PlotMacro.C`)
   * from a payload inspector application (see discussion at: https://github.com/cms-sw/cmssw/pull/19193#issuecomment-310304097 and following) 

The proposed implementation is actually inspired by David's https://github.com/cms-sw/cmssw/pull/19393#issuecomment-310397165, though I am not sure this is exactly what he had in mind at the time.

slava77 commented:
-1

merge conflicts

also, the hardcoded  pieces in LegacyTrackerTopology.h do not seem appropriate, given that there are ways to get the details from ES as it's done for all other tracker configurations.

mmusich commented on CalibTracker/SiStripLorentzAngle/src/LA_Filler.cc:
```diff
@@ -74,22 +75,25 @@ granularity(const SiStripDetId detid, const float tthetaL, const Long64_t TFE_in
+  if(localYbin_) gran += (localy < 0 ? "_yM":"_yP") + std::to_string(abs((int)(localy/localYbin_+(localy<0?-1:0))));
+  if(stripsPerBin_) gran += "_strip"+std::to_string((unsigned)((0.5+((apvstrip/64)?(127-apvstrip):apvstrip)/stripsPerBin_)*stripsPerBin_) );
   if(ensembleBins_) {
-    gran+= "_ensembleBin"+boost::lexical_cast<std::string>((int)(ensembleBins_*(tthetaL-ensembleLow_)/(ensembleUp_-ensembleLow_)));
```
@alesaggio, as for the merge conflicts, it's my bad:  #19645 has been merged first (though not fully signed). 
To avoid the conflict you can either revert commit 29cea77425adf94ed627ab3bbf6e7a3a3ca7734e, or rebase the branch following the instructions at http://cms-sw.github.io/tutorial-resolve-conflicts.html

davidlange6 commented:
> On Jul 21, 2017, at 9:21 AM, Marco Musich <notifications@github.com> wrote:
> 
> @mmusich commented on this pull request.
> 
> In DataFormats/TrackerCommon/interface/LegacyTrackerTopology.h:
> 
> > @@ -0,0 +1,98 @@
> +#ifndef TRACKER_TOPOLOGY_LEGACY_H
> +#define TRACKER_TOPOLOGY_LEGACY_H
> +
> +#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
> +#include "DataFormats/DetId/interface/DetId.h"
> +
> +#include <memory>
> +
> +// WARNING: this header has been introduced to call a TrackerTopology
> +// object whenever it is not possible to access it through an EventSetup.
> 
> Hi @slava77, a couple of use cases are:
> 
> 	• from within a ROOT macro (as in CalibTracker/SiStripChannelGain/test/UsefullCode/TimeHistory/PlotMacro.C)
> 	• from a payload inspector application (see discussion at: #19193 (comment) and following)
> The proposed implementation is actually inspired by David's #19393 (comment), though I am not sure this is exactly what he had in mind at the time.
> 

i personally don't see a better way except that all the hardwired numbers should get loaded from an existing xml (the path for which could get specified as an input argument?..)

that said, this PR seems to use the legacy data format in two places - one is a root macro - the other is a ESProducer, where I don't find its use valid. [but then again, I don't understand the design choice behind MeasureLA being an ESProducer... but maybe its obsolete in any case - it seems to be used only in a craft09 era cfg within cmssw..]

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
On 7/21/17 12:21 AM, Marco Musich wrote:
> Hi @slava77 <https://github.com/slava77>, a couple of use cases are:
> 
>   * from within a ROOT macro (as in
>     |CalibTracker/SiStripChannelGain/test/UsefullCode/TimeHistory/PlotMacro.C|)
>   * from a payload inspector application (see discussion at: #19193
>     (comment)
>     <https://github.com/cms-sw/cmssw/pull/19193#issuecomment-310304097>
>     and following)

Is there a solution for these two cases to handle phase-1 and phase-2
TrackerTopology use cases?
If the goal is to serve specifically SiStrip, perhaps this topology
should be called HardcodedSiStripTrackerTopology and should include
something unreadable for the pixel part in it.


@makortel
please remind what you have defined for the trackerNtuple case discussed
recently
in connection with #18491

makortel commented:
@slava77 
> please remind what you have defined for the trackerNtuple case discussed
> recently
> in connection with #18491

What I did in #18491 was that the ntuple has a branch for each of the DetId fields (or to be more precise, for each interesting function of TrackerTopology for a given subdet), which are then filled by calls to TrackerTopology (+ bunch of tricks to minimize copy-paste). Whether the phase0/1 or phase2 outer tracker fields are used is essentially dictated by a configuration flag that is set with eras.

franzoni commented:
hello is this PR still of relevance ?
if not, might it be closed ?

pieterdavid commented:
@franzoni: @alesaggio and I are working on an update that addresses the comments, so please do not close the PR yet

pieterdavid committed:
XML parsing for LegacyTrackerTopology

Renamed `LegacyTrackerTopology` to `StandaloneTrackerTopology`, and changed the signature of the method to `TrackerTopology fromTrackerParametersXML(const std::string&)`
The argument is the path of the `trackerParameters.xml` file to be parsed

cmsbuild commented:
Pull request #19393 was updated. @perrotta, @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @slava77, @davidlange6, @lpernie can you please check and sign again.

cmsbuild commented:
Pull request #19393 was updated. @perrotta, @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @slava77, @davidlange6, @lpernie can you please check and sign again.

alesaggio commented:
@pieterdavid changed the method to parse trackerParameters.xml.
I rebased to master to fix the merge conflicts

pieterdavid commented:
One more comment about `CalibTracker/SiStripLorentzAngle/plugins/MeasureLA.cc`: it's true that this is an ESProducer, but the method that does the work is called from the constructor (because with the example config the produce method isn't even called, I think - its implementation is essentially empty), where no EventSetup is available yet, so in fact this class looks like a ROOT macro that is initialized from a ParameterSet (maybe it would be a good idea if the people who are doing this calibration could have a look into changing this to something more appropriate than an ESProducer).

slava77 commented on DataFormats/TrackerCommon/BuildFile.xml:
```diff
@@ -1,8 +1,10 @@
+<use name="FWCore/Concurrency"/>
 <use name="FWCore/MessageLogger"/>
 <use name="FWCore/Utilities"/>
+<use name="xerces-c"/>
```
this dependency is not allowed

slava77 commented:
-1

dataformats can not depend on xerces

https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Restrictions_on_the_package_defi

pieterdavid commented:
I'm sorry, I did not think about that, I will move it somewhere else.
Would `CalibTracker/SiStripCommon` be a good place to put this?

pieterdavid committed:
Move StandaloneTrackerTopology to CalibTracker/SiStripCommon

cmsbuild commented:
Pull request #19393 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @davidlange6, @lpernie can you please check and sign again.

mmusich commented on CalibTracker/SiStripCommon/src/StandaloneTrackerTopology.cc:
```diff
@@ -0,0 +1,234 @@
+        }
+      }
+    }
+  } catch (...) {
```
@alesaggio @pieterdavid I think that the `catch(...)` construct is disallowed in CMSSW, at least it is flagged as violation of a code rule by the static analyzer.

pieterdavid replied:
Thanks for reviewing. I added an explicit catch for the remaining ones that xerces may throw

pieterdavid committed:
Remove catch(...) in StandaloneTrackerTopology.cc

cmsbuild commented:
Pull request #19393 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @davidlange6, @lpernie can you please check and sign again.

mmusich commented:
@arunhep @franzoni @lpernie can this PR be tested? Thanks

arunhep commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ad9acefb771a38ae28117bb8211954534f172aec
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19393/21895/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/8e59646c5dc2756e96963b11dceef23e96c92c1b
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19393/21895/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19393/21895/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19393/21895/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 2446803
 - DQMHistoTests: Total failures: 71143
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2375494
 - DQMHistoTests: Total skipped: 166
 - DQMHistoTests: Total Missing objects: 0
 - Checked 93 log files, 14 edm output root files, 23 DQM output files

arunhep commented:
+1

 . differences in HLT directory https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_3_X_2017-07-30-1100+19393/21650/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/HLT.html are probably because of this extra commit https://github.com/cms-sw/cmssw/commit/8e59646c5dc2756e96963b11dceef23e96c92c1b

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar

davidlange6 commented on CalibTracker/SiStripLorentzAngle/python/EnsembleCalibrationLA_cfg.py:
```diff
@@ -7,6 +7,8 @@
 
 from CalibTracker.SiStripLorentzAngle.MeasureLA_cff import METHOD_WIDTH, METHOD_PROB1, METHOD_AVGV2, METHOD_AVGV3, METHOD_RMSV2, METHOD_RMSV3
 
+process.load('Configuration.Geometry.GeometryIdeal_cff')
```
hi @alesaggio - you may want to confirm this is the geometry CFF that you want. It is usually the wrong one.

alesaggio replied:
Hi @davidlange6, how can I know which one is correct? I have no clue. Maybe @mmusich or @dimattia know?

mmusich replied:
Hi @alesaggio, I think that `Configuration.Geometry.GeometryRecoDB_cff` shall suffice for this use-case. We can follow to this one in another PR, the `CalibTracker/SiStripLorentzAngle` needs anyway heavy maintenance.

davidlange6 commented:
+1
